### DISCUSSION TITLE
Syntax of Opti stack to_function for parameter varying gradient.

### DISCUSSION THREAD

**Discussion by sixtosevenf** [casadi/casadi#3435] [Q&A] [answered]
Hi,

One cool thing with Opti stack is we can use `to_function`, making one-time function generation for optimization solver. Thus, I am trying to code an optimization solver with a specified gradient using Opti stack and `to_function`. 

The situation is as follows:
Let's assume the optimization problem is solved iteratively with respect to different gradients $G(x,`p')$. The gradient is coded with `casadi.Function` with the parameter $p$ and it is adopted to the solver using `opti.solver` command. The parameter $p$ will be submitted to the input of `to_function`. The below is a pseudo-code:

```
x = opti.variable();
p = opti.parameter();
cost = sumsqr(x);
grad_f = Function('grad_f',{x,p},{[],G(x,p)}); % Here, **the gradient function G is a function of optimization variable x and parameter p**
p_opts = struct('expand',1,'verbose',0,'grad_f',grad_f );
s_opts = struct('print_level',5);
opti.solver('ipopt',p_opts,s_opts)
opti.minimize(cost);
solver = opti.to_function('solver', {p}, {x}); % Here, I would like to the 'grad_f' is varying with respect to the parameter p
```

The goal here is one-time code generation without iteratively loading `grad_f = Function('grad_f',{x,p},{[],G(x,p)})` with respect to different G. I tried the above pseudo-code approach but the gradient is not changed with respect to the parameter when I simulated with `solver(p)`.

Does anyone know this type of approach (using casadi.Function and opti.to_function) or other ways for similar situations?
Any suggestion or comments would be appreciated.

**Comment by jgillis** [comment, marked as answer]
The problem is that p is not used in objective/constraints and hence discarded. A workaround could be to have `opti.minimize(cost+sumsqr(p))`. The absolute value of the objective is not important, so the solution should remain the same.

**Reply by sixtosevenf** [reply]
Thank Joris for your reply. The workaround works perfectly!
